### ISSUE TITLE
[Discussion] Is paddle.metric.Metric too minimal? A call for a more helpful base class

### ISSUE THREAD

**Issue by PlumBlossomMaid** [PaddlePaddle/Paddle#78100]
# [Discussion] Is paddle.metric.Metric too minimal? A call for a more helpful base class

# [讨论] paddle.metric.Metric 是否设计的过于简洁了？对更实用基类的呼吁

Hi PaddlePaddle Team,

Following the discussion in #78078 about `paddle.Model`, where the team acknowledged design issues in loss handling, I started examining another core high-level component: `paddle.metric.Metric`. This raises the question: does `Metric` suffer from similar "historical legacy" design choices?

If `paddle.Model` is like a "corrupt official" (redundant code, incomplete logging), then `paddle.metric.Metric` is more like a "laid-back cadre"—it hasn't done anything wrong, but it hasn't done much to help either. While not a bug, such minimalism can hinder developer productivity and project efficiency.

飞桨团队好，

继 #78078 关于 `paddle.Model` 的讨论之后（官方已确认 loss 处理上的设计问题），我开始关注另一个核心高层组件——`paddle.metric.Metric`。这引发了我的思考：`paddle.metric.Metric` 的设计是否也存在类似的“历史遗留”问题？

如果说 `paddle.Model` 是“违法乱纪”（代码冗余、功能缺失），那么 `paddle.metric.Metric` 更像是“躺平式干部”——什么都没做，但也没做错。躺平式干部虽没违法，但依旧需要治理，否则影响整体效能。

---

### 1. 真实案例：PaddleMaterials 的“用了如用”

In the PaddleMaterials project (`ppmat/metrics/diffnmr_metric.py`), I observed how `paddle.metric.Metric` is actually used:

**1.1 Inheritance Without Substance**
Although the project inherits from `paddle.metric.Metric`, the base class provides almost nothing:
- `reset` must be implemented manually
- `update` must be implemented manually
- `accumulate` must be implemented manually
- `name` must be implemented manually

**1.2 DIY Distributed Support**
The project had to write its own `_all_reduce_sum_` for multi-card reduction:
```python
def _all_reduce_sum_(t: paddle.Tensor) -> paddle.Tensor:
    if _is_dist():
        dist.all_reduce(t, op=dist.ReduceOp.SUM)
    return t
```

**1.3 Self-implemented `__call__` for Usability**
To simplify usage, ppmat even implemented its own `__call__` method.

**1.4 Code Duplication**
Every custom metric repeats the same `_sum`/`_n` pattern, with no unified state management like `add_state`.

**Conclusion**: PaddleMaterials' code demonstrates that the current `Metric` base class is "used but useless"—users bear the constraints of inheritance without receiving any substantial help. This vividly illustrates the "laid-back cadre" problem.

在 PaddleMaterials 项目（`ppmat/metrics/diffnmr_metric.py`）中，我观察到了 `paddle.metric.Metric` 的实际使用情况：

**1.1 继承但无实质帮助**
项目虽然继承了 `paddle.metric.Metric`，但基类几乎什么都没提供：
- `reset` 要自己实现
- `update` 要自己实现
- `accumulate` 要自己实现
- `name` 要自己实现

**1.2 分布式支持要自己造**
ppmat 不得不自己写 `_all_reduce_sum_` 来手动处理多卡归约：
```python
def _all_reduce_sum_(t: paddle.Tensor) -> paddle.Tensor:
    if _is_dist():
        dist.all_reduce(t, op=dist.ReduceOp.SUM)
    return t
```

**1.3 为简化使用自己封装 `__call__`**
为了能一步到位，ppmat 甚至自己实现了 `__call__` 方法。

**1.4 代码重复**
每个自定义指标都要重复写 `_sum`/`_n` 模式，没有 `add_state` 这样的统一管理机制。

**结论**：ppmat 的代码证明，现有 `Metric` 基类“用了如用”——用户既要承受继承的约束，又得不到任何实质帮助。这正是“躺平式干部”的生动体现。

---

### 2. “躺平”的具体表现

**2.1 The Base Class is Nearly Empty**

**2.1 基类几乎是空的**

```python
class Metric(metaclass=abc.ABCMeta):
    def __init__(self) -> None:
        pass
    
    @abc.abstractmethod
    def reset(self) -> None:
        raise NotImplementedError
    
    @abc.abstractmethod
    def update(self, *args: Any) -> None:
        raise NotImplementedError
    
    @abc.abstractmethod
    def accumulate(self) -> Any:
        raise NotImplementedError
    
    @abc.abstractmethod
    def name(self) -> str:
        raise NotImplementedError
    
    def compute(self, *args: Any) -> Any:
        return args
```

**2.2 State Management Entirely on Users**
- Each subclass manually defines state variables (e.g., `self.tp`, `self.fp`)
- Each subclass manually implements `reset`, prone to omission
- No unified registration mechanism like `add_state`

**2.3 Zero Distributed Support**
- Base class completely ignores multi-card scenarios
- ppmat had to implement `_all_reduce_sum_` itself
- Contrast with torchmetrics' declarative `dist_reduce_fx`

**2.4 Missing Device Management**
- In built-in metrics like `Accuracy`, Tensors are converted to numpy and moved back to CPU
- GPU acceleration is wasted
- No automatic synchronization in multi-card environments

**2.5 No Serialization Support**
- No `state_dict`/`load_state_dict` interface
- Metric states are lost when resuming from checkpoints

**2.6 No Metric Composition**
- Cannot combine multiple metrics into new ones
- Users have to rewrite new classes, repeating work

**2.7 Redundant Design: Three Steps vs One Step**
- Users must follow three steps: `compute` → `update` → `accumulate` [example](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/metric/Accuracy__upper_cn.html)
- Example code is 3x longer than torchmetrics [example](https://lightning.ai/docs/torchmetrics/stable/classification/accuracy.html)
- Contrast with torchmetrics' one-step `forward`

**2.2 状态管理完全推给用户**
- 每个子类都要手动定义状态变量（如 `self.tp`、`self.fp`）
- 每个子类都要手动实现 `reset`，容易遗漏
- 没有 `add_state` 这样的统一注册机制

**2.3 分布式支持为零**
- 基类完全不处理多卡场景
- ppmat 不得不自己实现 `_all_reduce_sum_`
- 对比 torchmetrics 的 `dist_reduce_fx` 声明式归约

**2.4 设备管理缺失**
- 在 `Accuracy` 等内置指标中，Tensor 被转成 numpy 回 CPU 计算
- GPU 加速白费
- 多卡环境下无法自动同步

**2.5 无法序列化保存**
- 没有 `state_dict`/`load_state_dict` 接口
- 断点续训时指标状态丢失

**2.6 不支持指标组合**
- 无法将多个指标组合成新指标
- 用户只能重写新类，重复造轮子

**2.7 设计冗余：三步走 vs 一步到位**
- 用户必须三步走：`compute` → `update` → `accumulate` [示例](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/metric/Accuracy__upper_cn.html)
- 示例代码比 torchmetrics 多 3 倍 [示例](https://lightning.ai/docs/torchmetrics/stable/classification/accuracy.html)
- 对比 torchmetrics 的 `forward` 一步搞定

---

### 3. 简洁不是错，但“简洁”和“简陋”有本质区别

Minimalism itself is not a problem—torchmetrics' base class maintains a clean interface while providing powerful infrastructure:
- Inherits from `nn.Module`, gaining device management and serialization
- `add_state` mechanism for centralized state management
- Base class implements `reset`, automatically resetting all registered states
- Declarative `dist_reduce_fx` for distributed reduction
- Supports `MetricCollection` and operator overloading

简洁本身不是问题，torchmetrics 的基类同样保持了接口简洁，但提供了强大的基础设施：
- 继承 `nn.Module`，自动获得设备管理、序列化能力
- `add_state` 机制集中管理状态
- 基类实现 `reset`，自动重置所有注册的状态
- `dist_reduce_fx` 声明式分布式归约
- 支持 `MetricCollection` 和运算符重载

---

### 4. 讨论方向：我们可以做什么？

1. **Could the base class inherit from `paddle.nn.Layer`?**
   - Gain automatic device management and serialization
   - Better integration with the Paddle ecosystem

2. **Could we introduce an `add_state` mechanism?**
   - Centralized state management
   - Automatic handling of `reset`, device transfer, and distributed reduction

3. **Could we build in distributed reduction support?**
   - Declarative `dist_reduce_fx` to specify reduction strategies
   - Free users from writing `_all_reduce_sum_` manually

4. **Could we unify the interface?**
   - Keep `update` and `compute` as core abstract methods
   - Provide `forward` as a unified entry point
   - Maintain backward compatibility by keeping `accumulate` as an alias with a DeprecationWarning

5. **Could we support metric composition?**
   - Implement `MetricCollection` for batch management
   - Support operator overloading for metric combinations

<p></p>

1. **基类能否继承 `paddle.nn.Layer`？**
   - 自动获得设备管理、序列化能力
   - 与 Paddle 生态更好融合

2. **能否引入 `add_state` 机制？**
   - 集中管理状态变量
   - 自动处理 `reset`、设备迁移、分布式归约

3. **能否内置分布式归约支持？**
   - 通过 `dist_reduce_fx` 声明式指定归约方式
   - 让用户不再手写 `_all_reduce_sum_`

4. **能否统一接口？**
   - 保留 `update` 和 `compute` 作为核心抽象方法
   - 提供 `forward` 作为统一入口
   - 兼容旧 API：保留 `accumulate` 作为别名，加 DeprecationWarning

5. **能否提供指标组合能力？**
   - 实现 `MetricCollection` 批量管理
   - 支持运算符重载，实现指标组合

---

### 5. 结语

As acknowledged in #78078, recognizing early design limitations is the first step towards improvement. I hope this discussion can pave the way for future enhancements to `paddle.metric.Metric`, ultimately making Paddle's high-level APIs more developer-friendly.

> *This issue is intended as a design discussion to prepare the ground for potential future contributions and refactoring. We are not yet at the implementation stage and look forward to hearing the community's and the team's thoughts on how to transform `paddle.metric.Metric` from a "laid-back cadre" into a truly helpful "doer".*

Thank you for your time and insights!

正如 #78078 中官方所回应的，承认早期设计未覆盖复杂场景是改进的第一步。我希望通过本次讨论，也能为 `paddle.metric.Metric` 的未来演进铺平道路。

> *本文仅作为设计讨论，旨在为未来可能的贡献和重构做思想准备。目前我们尚未到达具体实现阶段，希望能听到社区和官方的意见，共同探讨如何让 `paddle.metric.Metric` 从一个“躺平式干部”变成一个真正能帮助用户的“实干家”。*

期待您的回复！
